flocker/control/httpapi.py

```diff
@@ -212,6 +213,112 @@ def saved(ignored):
+        ]
+    )
+    @structured(
+        inputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        outputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        schema_store=SCHEMAS
+    )
+    def update_dataset(self, dataset_id, primary=None):
+        """
+        Update an existing dataset in the cluster configuration.
+
+        :param unicode dataset_id: The unique identifier of the dataset.  This
+            is a string giving a UUID (per RFC 4122).
+
+        :param unicode primary: The address of the node to which the dataset
+            will be moved.
+
+        :return: A ``dict`` describing the dataset which has been added to the
+            cluster configuration or giving error information if this is not
+            possible.
+        """
+        # Get the current configuration.
+        deployment = self.persistence_service.get()
+
+        # Find all manifestations of requested dataset.
+        manifestations_and_nodes = list(
+            other_manifestations_from_deployment(deployment, dataset_id)
+        )
+
+        # There are no manifestations containing the requested dataset.
+        if not manifestations_and_nodes:
+            raise DATASET_NOT_FOUND
+
+        # Lookup the node that has a primary Manifestation (if any)
+        primary_manifestation, origin_node = [
+            (manifestation, node)
+            for manifestation, node in manifestations_and_nodes
+            if manifestation.primary
+        ][0]
+
+        if origin_node.hostname == primary:
+            # Maybe return early here rather than bother the
+            # persistence_service.  raise Exception('New primary is the same as
+            # existing primary')
+            pass
+
+        # Now construct a new_deployment where the primary manifestation of the
+        # dataset is on the requested primary node.
+        new_origin_node = Node(
+            hostname=origin_node.hostname,
+            applications=origin_node.applications,
+            other_manifestations=(
+                origin_node.other_manifestations
+                - frozenset({primary_manifestation})
+            )
+        )
+        deployment = deployment.update_node(new_origin_node)
+
+        primary_nodes = list(
+            node for node in deployment.nodes if primary == node.hostname
+        )
+        if len(primary_nodes) == 0:
+            # `primary` is not in cluster. Add it.
+            # XXX Check cluster state to determine if the given primary node
+            # actually exists.  If not, raise PRIMARY_NODE_NOT_FOUND.
+            # See FLOC-1278
+            new_target_node = Node(
+                hostname=primary,
+                other_manifestations=frozenset({primary_manifestation})
+            )
+        else:
+            # There should only be one node with the requested primary
+            # hostname. ``IndexError`` here if that's not the case.
+            (target_node,) = primary_nodes
```

Comment:
Changed the Exception in the comment, but not sure what the advantage of list unpacking is. Just to avoid having to add the trailing comma?